### DISCUSSION TITLE
Can single_source_bellman_ford handle MultiDiGraphs?

### DISCUSSION THREAD

**Discussion by demogorgi** [networkx/networkx#5110] [Q&A]
I have a `MultiDiGraph` with some negative edge weights. When I call the `negative_edge_cycle`-method on it, it returns `False`. However during `single_source_bellman_ford` a `NetworkXUnbounded`-exception is raised: `networkx.exception.NetworkXUnbounded: Negative cost cycle detected`.

**Comment by dschult** [comment]
The `negative_edge_cycle` makes the graph a connected graph by adding a node connected to all other nodes with edges that have weight 1, then it runs `bellman_ford_predecessor_and_distance` starting from that fictitious node and checks for whether that raises that very exception. If `negative_edge_cycle` is returning `False`, then the bellman_ford algorithm is not raising the exception.  If `single_source_bellman_ford` is raising the exception then the bellman_ford algorithm is raising an exception. The addition of that extra node (with no incoming edges) will not affect whether cycles exist or not.

So, what you are reporting should not be happening.  Can you give a reasonably small example where this happens?  It sounds like your code is actually doing these two things on different graphs.

**Reply by demogorgi** [reply]
In terms of code I have the following small example:
```
import networkx as nx

print("\nUsing networkx version {}".format(nx.__version__))

s, t = ['1688', '35175']

# !!! On Windows !!!
# With networkx 2.4, the method 'single_source_bellman_ford' seems to work on the provided test data.
# With networkx 2.5, 2.5.1, 2.6, 2.6.1 and 2.6.2 a negative cost cycle is wrongly detected.

G = nx.read_weighted_edgelist("test.weighted.edgelist", create_using=nx.MultiDiGraph())
print("Negative cycle? {}".format(nx.negative_edge_cycle(G)))
es = G.edges(data=True)
ns = G.nodes(data=True)
print("Graph has {} nodes and {} edges.".format(len(es), len(ns)))
print("Computing shortest path from {} to {}.".format(s, t))
length, path = nx.single_source_bellman_ford(G, s, t, weight='weight')

print("\n\nlength: {}\n".format(length))
print("path: {}".format(path))
```
Unfortunately the file test.weighted.edgelist is almost 10 MB. I zipped it to 3 MB and it can be downloaded from https://www.dropbox.com/s/kysmos6lvwzn8f9/test.weighted.edgelist.zip?dl=0

Please note that I only observe the described behaviour on windows and that the code seems to work for networkx 2.4 but not for higher versions.

**Comment by dschult** [comment]
I can verify that the example graph from above leads to a case where `negative_edge_cycle` returns `False` and yet an error is raised by the bellman ford routines.  I am trying to identify what might be the cause using ideas/code from #5089. I need a smaller example to be able to see what is happening. I can see that ~112K nodes are explored before raising the error. Trying to get a smaller example that does the same thing.
